File: reverse-arr-in-place/JS/reverseArrInPlace.js
// Reverse an array in-place.

// Note: reversing in-place means reversing the original list.

// --
// Arrays for testing
// --

const arr = [1, 2, 3, 4, 5];

// while min index and max index are different we keep moving the to the center

function reverseArr(arr) {
    if (!Array.isArray(arr) || arr.length === 0 || !arr.every(e => typeof(e) === 'number')) {
        console.error('error: invalid array');
        return undefined;
    }
    
    let minIndex = 0;
    let maxIndex = arr.length - 1;
    
    while (minIndex < maxIndex) {
        const temp = arr[minIndex];
        arr[minIndex] = arr[maxIndex];
        arr[maxIndex] = temp;
        minIndex++;
        maxIndex--;
    }
    return true;
}

let result = reverseArr(arr);
if (result) console.log(arr);

File: reverse-arr-in-place/C/reverse-arr-in-place.c
// Reverse an array in-place.

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define ARR_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))

bool reverse_arr(int *arr, size_t arr_size) 
{
    if (arr == NULL || arr_size == 0) {
        fprintf(stderr, "error: invalid array\n");
        return false;
    }
    
    int min_index = 0;
    int max_index = arr_size - 1;

    while (min_index < max_index) {
        int tmp = arr[min_index];
        arr[min_index] = arr[max_index];
        arr[max_index] = tmp;

        min_index++;
        max_index--;
    }
    return true;
}

void print_arr(int *arr, size_t arr_size) 
{
    for (size_t i = 0; i < arr_size; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main(void)
{
    // array for testing:
    int arr[] = {1, 2, 3, 4, 5};

    if (reverse_arr(arr, ARR_SIZE(arr))) {
        print_arr(arr, ARR_SIZE(arr));
    }
    return EXIT_SUCCESS;
}
